ssl/statem/statem_lib.c

```diff
@@ -624,15 +624,29 @@ CON_FUNC_RETURN tls_construct_finished(SSL_CONNECTION *s, WPACKET *pkt)
      * moment. If we didn't already do this when we sent the client certificate
      * then we need to do it now.
      */
-    if (SSL_CONNECTION_IS_TLS13(s)
-            && !s->server
-            && (s->early_data_state != SSL_EARLY_DATA_NONE
-                || (s->options & SSL_OP_ENABLE_MIDDLEBOX_COMPAT) != 0)
+    if (SSL_IS_QUIC_HANDSHAKE(s)) {
+        if (!s->server
+            && s->early_data_state != SSL_EARLY_DATA_NONE
             && s->s3.tmp.cert_req == 0
-            && (!ssl->method->ssl3_enc->change_cipher_state(s,
-                    SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_WRITE))) {;
-        /* SSLfatal() already called */
-        return CON_FUNC_ERROR;
+            && s->write_key_yielded == 0) {
+                if (!ssl->method->ssl3_enc->change_cipher_state(s,
+                        SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_WRITE)) {
+                    /* SSLfatal() already called */
+                    return CON_FUNC_ERROR;
+                }
+                s->write_key_yielded = 1;
+        }
+    } else {
```

Comment:
I would like to have the branch for TLS13 consistent with QUIC above so something like this
```
} else if (SSL_CONNECTION_IS_TLS13(s)) {
    if (!s->server....
```
would be better in my opinion.